ndb/src/google/cloud/ndb/query.py

```diff
@@ -129,43 +140,729 @@ def resolve(self, bindings, used):
+                #        [Sn and A1 and ... and Am]
+                for segment in self.or_parts:
+                    # ``segment`` represents ``Si``
+                    segment.extend(node._nodes)
+            else:
+                #    [S1 or ... or Sn] and [A1]
+                # -> [S1 and A1] or ... or [Sn and A1]
+                for segment in self.or_parts:
+                    segment.append(node)
 
 
 class ConjunctionNode(Node):
-    def __init__(self, *args, **kwargs):
-        raise NotImplementedError
+    """Tree node representing a boolean ``AND`` operator on multiple nodes.
+
+    .. warning::
+
+        The constructor for this type may not always return a
+        :class:`ConjunctionNode`. For example:
+
+        * If the passed in ``nodes`` has only one entry, that single node
+          will be returned by the constructor
+        * If the resulting boolean expression has an ``OR`` in it, then a
+          :class:`DisjunctionNode` will be returned; e.g.
+          ``AND(OR(A, B), C)`` becomes ``OR(AND(A, C), AND(B, C))``
+
+    Args:
+        nodes (Tuple[Node, ...]): A list of nodes to be joined.
+
+    Raises:
+        TypeError: If ``nodes`` is empty.
+        RuntimeError: If the ``nodes`` combine to an "empty" boolean
+            expression.
+    """
+
+    __slots__ = ("_nodes",)
+
+    def __new__(cls, *nodes):
+        if not nodes:
+            raise TypeError("ConjunctionNode() requires at least one node.")
+        elif len(nodes) == 1:
+            return nodes[0]
+
+        clauses = _BooleanClauses("ConjunctionNode", combine_or=False)
+        for node in nodes:
+            clauses.add_node(node)
+
+        if not clauses.or_parts:
+            # NOTE: The original implementation returned a ``FalseNode``
+            #       here but as far as I can tell this code is unreachable.
+            raise RuntimeError("Invalid boolean expression")
+
+        if len(clauses.or_parts) > 1:
+            return DisjunctionNode(
+                *[ConjunctionNode(*segment) for segment in clauses.or_parts]
+            )
+
+        self = super(ConjunctionNode, cls).__new__(cls)
+        self._nodes = clauses.or_parts[0]
+        return self
+
+    def __getnewargs__(self):
+        """Private API used to specify ``__new__`` arguments when unpickling.
+
+        .. note::
+
+            This method only applies if the ``pickle`` protocol is 2 or
+            greater.
+
+        Returns:
+            Tuple[Node, ...]: The list of stored nodes, converted to a
+            :class:`tuple`.
+        """
+        return tuple(self._nodes)
+
+    def __iter__(self):
+        return iter(self._nodes)
+
+    def __repr__(self):
+        all_nodes = ", ".join(map(str, self._nodes))
+        return "AND({})".format(all_nodes)
+
+    def __eq__(self, other):
+        if not isinstance(other, ConjunctionNode):
+            return NotImplemented
+
+        return self._nodes == other._nodes
```

Comment:
`__hash__`?

Reply:
I'm not sure what the question is, could you give more details?

Reply:
Pretty much any time I see an `__eq__` method, I look for `__hash__` as well, for immutables at least.

Reply:
Would you like me to implement it? The default will work as well?

Reply:
Maybe it is YAGNI, if application code doesn't hold on to node instances directly.  I guess we know that the library doesn't put nodes into sets or use them as dict keys?

Reply:
Correct. Given `ndb` is 7+ years old, it seems all "new" behavior is YAGNI?